File: app/routes/chatbot.py
from flask import Blueprint, request, jsonify
from app.services.chatbot_service import ChatbotService
import logging

chatbot_bp = Blueprint('chatbot', __name__, url_prefix='/api/chatbot')
logger = logging.getLogger(__name__)


# ============================================================
# ENDPOINT 1: START OR RESUME SESSION
# ============================================================

@chatbot_bp.route('/start', methods=['POST'])
def start_session():
    """
    Start or resume a chatbot session
    
    Expected JSON: {
        "user_id": 1,
        "language": "en"  // optional, default is "en"
    }
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'success': False, 'message': 'No data provided'}), 400
            
        user_id = data.get('user_id')
        if not user_id:
            return jsonify({'success': False, 'message': 'user_id is required'}), 400
            
        language = data.get('language', 'en').lower()
        
        result = ChatbotService.start_session(user_id=user_id, language=language)
        
        if result.get('success'):
            return jsonify(result), 200
        else:
            return jsonify(result), 400
            
    except Exception as e:
        logger.error(f"Chatbot start session error: {str(e)}")
        return jsonify({'success': False, 'message': f'Error starting session: {str(e)}'}), 500


# ============================================================
# ENDPOINT 2: SEND MESSAGE
# ============================================================

@chatbot_bp.route('/message', methods=['POST'])
def send_message():
    """
    Process a user message and return the assistant reply
    
    Expected JSON: {
        "user_id": 1,
        "session_id": 12,
        "message": "I have a headache"
    }
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'success': False, 'message': 'No data provided'}), 400
            
        user_id = data.get('user_id')
        session_id = data.get('session_id')
        user_message = data.get('message')
        
        if not user_id or not session_id or not user_message:
            return jsonify({'success': False, 'message': 'user_id, session_id, and message are required'}), 400
            
        result = ChatbotService.send_message(user_id=user_id, session_id=session_id, user_text=user_message)
        
        if result.get('success'):
            return jsonify(result), 200
        else:
            return jsonify(result), 400
            
    except Exception as e:
        logger.error(f"Chatbot send message error: {str(e)}")
        return jsonify({'success': False, 'message': f'Error sending message: {str(e)}'}), 500


# ============================================================
# ENDPOINT 3: RESET SESSION
# ============================================================

@chatbot_bp.route('/reset', methods=['POST'])
def reset_session():
    """
    Archive the current session and open a fresh one.
    
    Expected JSON: {
        "user_id": 1,
        "session_id": 12,
        "language": "en"  // optional
    }
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'success': False, 'message': 'No data provided'}), 400
            
        user_id = data.get('user_id')
        session_id = data.get('session_id')
        language = data.get('language', 'en').lower()
        
        if not user_id or not session_id:
            return jsonify({'success': False, 'message': 'user_id and session_id are required'}), 400
            
        result = ChatbotService.reset_session(user_id=user_id, session_id=session_id, language=language)
        
        if result.get('success'):
            return jsonify(result), 200
        else:
            return jsonify(result), 400
            
    except Exception as e:
        logger.error(f"Chatbot reset session error: {str(e)}")
        return jsonify({'success': False, 'message': f'Error resetting session: {str(e)}'}), 500


# ============================================================
# ENDPOINT 4: GET HISTORY
# ============================================================

@chatbot_bp.route('/history/<int:user_id>', methods=['GET'])
def get_history(user_id):
    """
    Return all sessions and messages for a user
    """
    try:
        result = ChatbotService.get_history(user_id=user_id)
        
        if result.get('success'):
            return jsonify(result), 200
        else:
            return jsonify(result), 400
            
    except Exception as e:
        logger.error(f"Chatbot get history error: {str(e)}")
        return jsonify({'success': False, 'message': f'Error getting history: {str(e)}'}), 500

File: app/static/js/chatbot.js
/**
 * chatbot.js
 * Handles chatbot session management, messaging, and emergency flag.
 *
 * Depends on: api.js, translation.js, auth.js
 * Exposes: global `Chatbot` object
 *
 * Usage in chatbot.html {% block scripts %}:
 *   await Chatbot.init(userId, language, memberName);
 *   await Chatbot.sendMessage(text);
 *   Chatbot.reset();
 */

const Chatbot = (function () {

    // ─────────────────────────────────────────
    // STATE
    // ─────────────────────────────────────────
    let _userId = null;
    let _sessionId = null;
    let _language = 'en';
    let _memberName = '';
    let _isLoading = false;

    // Callbacks set by the page
    let _onMessage = null;   // fn({ role, content, isEmergency })
    let _onLoading = null;   // fn(bool)
    let _onEmergency = null;   // fn(message)
    let _onError = null;   // fn(message)


    // ─────────────────────────────────────────
    // PRIVATE: set loading state
    // ─────────────────────────────────────────
    function _setLoading(state) {
        _isLoading = state;
        if (typeof _onLoading === 'function') _onLoading(state);
    }


    // ─────────────────────────────────────────
    // PRIVATE: trigger emergency
    // ─────────────────────────────────────────
    function _triggerEmergency(replyText) {
        const msg = t('chatbot.emergency_msg');
        // Show global banner from base.html
        if (typeof showEmergency === 'function') showEmergency(msg);
        // Also call page-level handler if set
        if (typeof _onEmergency === 'function') _onEmergency(msg);
    }


    // ─────────────────────────────────────────
    // PUBLIC: on(event, callback)
    // Register page-level event handlers.
    // Events: 'message', 'loading', 'emergency', 'error'
    // ─────────────────────────────────────────
    function on(event, callback) {
        if (event === 'message') _onMessage = callback;
        if (event === 'loading') _onLoading = callback;
        if (event === 'emergency') _onEmergency = callback;
        if (event === 'error') _onError = callback;
    }


    // ─────────────────────────────────────────
    // PUBLIC: init(userId, language, memberName)
    // Start or resume a session.
    // Calls /api/chatbot/start and stores session_id.
    // Returns { success, sessionId } or { success: false }
    // ─────────────────────────────────────────
    async function init(userId, language, memberName) {
        _userId = userId;
        _language = language || getCurrentLang() || 'en';
        _memberName = memberName || '';

        _setLoading(true);

        const result = await API.post('/api/chatbot/start', {
            user_id: _userId,
            language: _language,
        });

        _setLoading(false);

        if (!result.success) {
            console.error('[Chatbot.init]', result.message);
            if (typeof _onError === 'function') _onError(result.message || t('common.error'));
            return { success: false };
        }

        _sessionId = result.session_id;

        // Emit the first question from the backend (opening prompt for current stage)
        const firstQ = result.first_question || result.message || '';
        if (firstQ && typeof _onMessage === 'function') {
            _onMessage({ role: 'assistant', content: firstQ, isEmergency: false });
        }

        return { success: true, sessionId: _sessionId };
    }


    // ─────────────────────────────────────────
    // PUBLIC: sendMessage(text)
    // Send a user message and receive assistant reply.
    // Emits via _onMessage callback.
    // Returns { success, reply, isEmergency } or { success: false }
    // ─────────────────────────────────────────
    async function sendMessage(text) {
        if (!_sessionId) {
            console.error('[Chatbot.sendMessage] No active session.');
            if (typeof _onError === 'function') _onError(t('common.error'));
            return { success: false };
        }

        if (_isLoading) return { success: false };

        const trimmed = text.trim();
        if (!trimmed) return { success: false };

        // Emit user message immediately (optimistic UI)
        if (typeof _onMessage === 'function') {
            _onMessage({ role: 'user', content: trimmed, isEmergency: false });
        }

        _setLoading(true);

        const result = await API.post('/api/chatbot/message', {
            user_id: _userId,
            session_id: _sessionId,
            message: trimmed,
        });

        _setLoading(false);

        if (!result.success) {
            console.error('[Chatbot.sendMessage]', result.message);
            if (typeof _onError === 'function') _onError(result.message || t('common.error'));
            return { success: false };
        }

        const reply = result.reply || '';
        const isEmergency = result.is_emergency === true;

        // Emit assistant reply
        if (typeof _onMessage === 'function') {
            _onMessage({ role: 'assistant', content: reply, isEmergency });
        }

        // Trigger emergency flow if flagged
        if (isEmergency) _triggerEmergency(reply);

        return { success: true, reply, isEmergency };
    }


    // ─────────────────────────────────────────
    // PUBLIC: reset(language?)
    // Archive current session and open a fresh one.
    // Returns { success, sessionId } or { success: false }
    // ─────────────────────────────────────────
    async function reset(language) {
        if (!_sessionId) return { success: false };

        _setLoading(true);

        const result = await API.post('/api/chatbot/reset', {
            user_id: _userId,
            session_id: _sessionId,
            language: language || _language,
        });

        _setLoading(false);

        if (!result.success) {
            console.error('[Chatbot.reset]', result.message);
            if (typeof _onError === 'function') _onError(result.message || t('common.error'));
            return { success: false };
        }

        _sessionId = result.session_id;

        // Hide emergency banner if visible
        if (typeof hideEmergency === 'function') hideEmergency();

        return { success: true, sessionId: _sessionId };
    }


    // ─────────────────────────────────────────
    // PUBLIC: getHistory(userId)
    // Fetch all past sessions for a user.
    // Returns { success, sessions } or { success: false }
    // ─────────────────────────────────────────
    async function getHistory(userId) {
        const result = await API.get(`/api/chatbot/history/${userId || _userId}`);

        if (!result.success) {
            console.error('[Chatbot.getHistory]', result.message);
            return { success: false, sessions: [] };
        }

        return { success: true, sessions: result.history || [] };
    }


    // ─────────────────────────────────────────
    // PUBLIC: getState()
    // Returns current session state (useful for debugging)
    // ─────────────────────────────────────────
    function getState() {
        return {
            userId: _userId,
            sessionId: _sessionId,
            language: _language,
            memberName: _memberName,
            isLoading: _isLoading,
        };
    }


    // ─────────────────────────────────────────
    // EXPOSE
    // ─────────────────────────────────────────
    return {
        on,
        init,
        sendMessage,
        reset,
        getHistory,
        getState,
    };

})();